### PR TITLE
61_吴昊_第三课作业

### PR DESCRIPTION
**Pull request by CN-AlbertWu96**
@zyzlouis 
对于第二题我的理解是修改address同时mapping的key也需要修改，因为employeeId在该和合约中和employees[employeeId].id(即account的address)一致，所以没有一个id可以指明employee身份。

### CODE REVIEW

File: Lesson3/assignment/multiPayRoll_new.sol [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+- contract C is O
+- contract K1 is A, B
+- contract K2 is A, C
+- contract Z is K1, K2
+*/
+
+第一题：
+
+截图在'./img'下。
+
```
Comment (zyzlouis) [inline Lesson3/assignment/multiPayRoll_new.sol]: 截图少了一张，你总共有7个函数。另外建议下次有提交截图类作业的时候，截取函数调用部分就可以了，不用截取其他的内容，图片看上去更直观一些。

File: Lesson3/assignment/multiPayRoll_new.sol [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+function changePaymentAddress(address originId, address newId) employeeExist(originId) employeeNotExist(newId) public{
+    var originEmployee = employees[originId];
+    _partialPaid(originEmployee);
+    employees[newId] = Employee(newId, originEmployee.salary, now);
+    delete employees[originId];
+}
+
+第三题（加分题）：
+
```
Comment (zyzlouis) [inline Lesson3/assignment/multiPayRoll_new.sol]: C3 linearization中，有一个继承顺序你理解反了：
比如 contract K1 is A, B {...} 的意思是K1继承自B和A，同时B是离K1最近的继承关系(most-derived), 所以：

L(O) = [O]
L(A) = [A, O]
L(B) = [B, O]
L(C) = [C, O]
L(K2) = [K2] + merge(L(C), L(A), [C, A])
= [K2] + merge([C, O], [A, O], [C, A])
= [K2, C] + merge([O], [A, O], [A])
= [K2, C, A, O]
L(K1) = [K1] + merge(L(B), L(A), [B, A])
= [K1] + merge([B, O], [A, O], [B, A])
= [K1, B] + merge([O], [A, O], [A])
= [K1, B, A, O]

L(Z) = [Z] + merge(L(K2), L(K1), [K2, K1])
= [Z] + merge([K2, C, A, O], [K1, B, A, O], [K2, K1])
= [Z, K2] + merge([C, A, O], [K1, B, A, O], [K1])
= [Z, K2, C] + merge([A, O], [K1, B, A, O], [K1])
= [Z, K2, C, K1] + merge([A, O], [B, A, O])
= [Z, K2, C, K1, B] + merge([A, O], [A, O])
= [Z, K2, C, K1, B, A, O]

File: Lesson3/assignment/multiPayRoll_new.sol [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        assert(nextPayday <= now);
+        employee.lastPayday = now;
+        employee.id.transfer(employee.salary);
+    }
+}
+
+第二题：
+
+确保原有id存在并且改变后id不存在。
+
```
Comment (zyzlouis) [inline Lesson3/assignment/multiPayRoll_new.sol]: 完成的不错

File: Lesson3/assignment/multiPayRoll_new.sol (hunk 1 above)
Comment (CN-AlbertWu96) [reply Lesson3/assignment/multiPayRoll_new.sol]: changepaymentaddress

File: Lesson3/assignment/multiPayRoll_new.sol (hunk 2 above)
Comment (CN-AlbertWu96) [reply Lesson3/assignment/multiPayRoll_new.sol]: 恩恩好的，我明白了！多谢！

**Pull request merged**